compiler/rustc_resolve/src/def_collector.rs

```diff
@@ -363,27 +369,61 @@ impl<'a, 'ra, 'tcx> visit::Visitor<'a> for DefCollector<'a, 'ra, 'tcx> {
         if let MgcaDisambiguation::Direct = constant.mgca_disambiguation
             && self.resolver.tcx.features().min_generic_const_args()
         {
-            visit::walk_anon_const(self, constant);
+            self.with_direct_const_arg(true, |this| {
+                visit::walk_anon_const(this, constant);
+            });
             return;
         }
-
-        let parent = self.create_def(constant.id, None, DefKind::AnonConst, constant.value.span);
-        self.with_parent(parent, |this| visit::walk_anon_const(this, constant));
+        
+        self.with_direct_const_arg(false, |this| {
+            let parent =
+                this.create_def(constant.id, None, DefKind::AnonConst, constant.value.span);
+            this.with_parent(parent, |this| visit::walk_anon_const(this, constant));
+        })
     }
 
     fn visit_expr(&mut self, expr: &'a Expr) {
+        let handle_const_block = |this: &mut Self, constant: &'a AnonConst, def_kind: DefKind| {
+            for attr in &expr.attrs {
+                visit::walk_attribute(this, attr);
+            }
+
+            let def = this.create_def(constant.id, None, def_kind, constant.value.span);
+            this.with_direct_const_arg(false, |this| {
+                this.with_parent(def, |this| visit::walk_anon_const(this, constant));
+            });
+        };
+
         let parent_def = match expr.kind {
             ExprKind::MacCall(..) => return self.visit_macro_invoc(expr.id),
             ExprKind::Closure(..) | ExprKind::Gen(..) => {
                 self.create_def(expr.id, None, DefKind::Closure, expr.span)
             }
             ExprKind::ConstBlock(ref constant) => {
-                for attr in &expr.attrs {
-                    visit::walk_attribute(self, attr);
+                handle_const_block(self, constant, DefKind::InlineConst);
+                return;
+            }
+            ExprKind::Struct(ref se) if self.invocation_parent.in_direct_const_arg => {
+                let StructExpr { qself, path, fields, rest } = &**se;
+
+                for init_expr in fields {
+                    if let ExprKind::ConstBlock(ref constant) = init_expr.expr.kind {
+                        handle_const_block(self, constant, DefKind::AnonConst);
```

Comment:
please add a comment that we only do this to give const blocks a different `DefKind`, the rest is just a `walk`